File: imagepipeline/src/test/java/com/facebook/imagepipeline/memory/PooledByteStreamsTest.java
/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.facebook.imagepipeline.memory;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;

import org.robolectric.RobolectricTestRunner;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class PooledByteStreamsTest  {
  private static final int POOLED_ARRAY_SIZE = 4;

  private ByteArrayPool mByteArrayPool;
  private byte[] mPooledArray;

  private byte[] mData;
  private InputStream mIs;
  private ByteArrayOutputStream mOs;

  private PooledByteStreams mPooledByteStreams;

  @Before
  public void setUp() {
    mByteArrayPool = mock(ByteArrayPool.class);
    mData = new byte[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 115};
    mIs = new ByteArrayInputStream(mData);
    mOs = new ByteArrayOutputStream();

    mPooledArray = new byte[4];
    mPooledByteStreams = new PooledByteStreams(mByteArrayPool, POOLED_ARRAY_SIZE);
    when(mByteArrayPool.get(POOLED_ARRAY_SIZE)).thenReturn(mPooledArray);
  }

  @Test
  public void testUsesPool() throws IOException {
    mPooledByteStreams.copy(mIs, mOs);
    verify(mByteArrayPool).get(POOLED_ARRAY_SIZE);
    verify(mByteArrayPool).release(mPooledArray);
  }

  @Test
  public void testReleasesOnException() throws IOException {
    try {
      mPooledByteStreams.copy(
          mIs,
          new OutputStream() {
            @Override
            public void write(int oneByte) throws IOException {
              throw new IOException();
            }
          });
      fail();
    } catch (IOException ioe) {
      // expected
    }

    verify(mByteArrayPool).release(mPooledArray);
  }

  @Test
  public void testCopiesData() throws IOException {
    mPooledByteStreams.copy(mIs, mOs);
    assertArrayEquals(mData, mOs.toByteArray());
  }

  @Test
  public void testReleasesOnExceptionWithSize() throws IOException {
    try {
      mPooledByteStreams.copy(
          mIs,
          new OutputStream() {
            @Override
            public void write(int oneByte) throws IOException {
              throw new IOException();
            }
          }, 3);
      fail();
    } catch (IOException ioe) {
      // expected
    }

    verify(mByteArrayPool).release(mPooledArray);
  }

  @Test
  public void testCopiesDataWithSize() throws IOException {
    mPooledByteStreams.copy(mIs, mOs, 3);
    assertArrayEquals(Arrays.copyOf(mData, 3), mOs.toByteArray());
  }
}

File: imagepipeline/src/test/java/com/facebook/common/memory/PooledByteStreamsTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.common.memory

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.Arrays
import org.assertj.core.api.Assertions
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class PooledByteStreamsTest {
  companion object {
    private const val POOLED_ARRAY_SIZE = 4
  }

  private lateinit var byteArrayPool: ByteArrayPool
  private lateinit var pooledArray: ByteArray

  private lateinit var data: ByteArray
  private lateinit var inputStream: InputStream
  private lateinit var outputStream: ByteArrayOutputStream

  private lateinit var pooledByteStreams: PooledByteStreams

  @Before
  fun setUp() {
    byteArrayPool = Mockito.mock(ByteArrayPool::class.java)
    data = byteArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 115)
    inputStream = ByteArrayInputStream(data)
    outputStream = ByteArrayOutputStream()

    pooledArray = ByteArray(4)
    pooledByteStreams = PooledByteStreams(byteArrayPool, POOLED_ARRAY_SIZE)
    Mockito.`when`(byteArrayPool.get(POOLED_ARRAY_SIZE)).thenReturn(pooledArray)
  }

  @Test
  @Throws(IOException::class)
  fun testUsesPool() {
    pooledByteStreams.copy(inputStream, outputStream)
    Mockito.verify(byteArrayPool).get(POOLED_ARRAY_SIZE)
    Mockito.verify(byteArrayPool).release(pooledArray)
  }

  @Test
  @Throws(IOException::class)
  fun testReleasesOnException() {
    try {
      pooledByteStreams.copy(
          inputStream,
          object : OutputStream() {
            @Throws(IOException::class)
            override fun write(oneByte: Int) {
              throw IOException()
            }
          })
      Assertions.fail("Expected IOException was not thrown")
    } catch (ioe: IOException) {
      // expected
    }

    Mockito.verify(byteArrayPool).release(pooledArray)
  }

  @Test
  @Throws(IOException::class)
  fun testCopiesData() {
    pooledByteStreams.copy(inputStream, outputStream)
    Assertions.assertThat(outputStream.toByteArray()).isEqualTo(data)
  }

  @Test
  @Throws(IOException::class)
  fun testReleasesOnExceptionWithSize() {
    try {
      pooledByteStreams.copy(
          inputStream,
          object : OutputStream() {
            @Throws(IOException::class)
            override fun write(oneByte: Int) {
              throw IOException()
            }
          },
          3)
      Assertions.fail("Expected IOException was not thrown")
    } catch (ioe: IOException) {
      // expected
    }

    Mockito.verify(byteArrayPool).release(pooledArray)
  }

  @Test
  @Throws(IOException::class)
  fun testCopiesDataWithSize() {
    pooledByteStreams.copy(inputStream, outputStream, 3)
    Assertions.assertThat(outputStream.toByteArray()).isEqualTo(Arrays.copyOf(data, 3))
  }
}
